proposals/0000-linear-types.rst

```diff
@@ -0,0 +1,858 @@
+opportunity is easily detected and the former definition of ``g`` is
+understood as the latter, well-typed, one. It means that is not a
+breaking change to strengthen a *first-order* regular arrow ``->``
+into a linear ``->.`` in an interface.
+
+An important point to note is that ``case_0`` is meaningless: it makes
+it possible to create values dependending on a value which may not
+exist. For instance the length of a list argument with multiplicity
+``0``.
+
+::
+
+  -- Wrong!
+  badLength :: [a] :'0-> Int
+  badLength [] = 0
+  badLength (_:l) = 1 + badLength l
+
+  -- Not linear! But well-typed if the above is accepted
+  f :: [a] ->. (Int, [a])
+  f l = (badLength l, l)
+
+Because we want to allow ``case_p`` for a variable ``p``, This
+creates a small complication where variables never stand for ``0``, in
+particular type-application of multiplicity variables must prohibit
+``0``.
+
+There are unresolved issue regarding inference (see Unresolved
+questions below for a more precise description):
+
+- There is no account of multiplicity inference. A better
+  understanding would make inference more predictable.
+- For ``let`` bindings and ``case`` expressions which are not part of
+  an equation, we want to infer the multiplicity annotation. The
+  process for this is not yet defined.
+
+Effect and Interactions
+-----------------------
+
+A staple of this proposal is that it does not modify Haskell for those
+who don't want to use it, or don't know of linear types. Even if an
+API exports linear types, they are easy to ignore: just imagine that
+the arrows are regular arrows, it will work as expected.
+
+Linear data types are just regular Haskell type, which means its cheap
+to interact with existing libraries. That is, unless there are linear
+arrows in argument position. In which case, attempt to use a
+non-linear function will raise a linear-type error. The motivating
+examples are all like this: they are libraries which require linear
+types to work.
+
+There is an unpleasant interaction with ``-XRebindableSyntax``: ``if u
+then t else e`` is interpreted as ``ifThenElse u t e``. Unfortunately,
+these two construct have differrent typing rules when ``t`` and ``e``
+have free linear variables. Therefore well-typed linearly typed
+programs can stop typing when ``-XRebindableSyntax`` is added.
+
+The meta-theory of linear types in a lazy language fail if we allow
+unrestricted ``newtype``-s:
+
+::
+
+  newtype Unrestricted' a where
+    Unrestricted' :: a -> Unrestricted' a
+
+Intuitively, this is because forcing a value ``v :: Unrestricted a``
+has the consequence of consuming all the resources in the closure of
+``v`` making it safe to use the value many times or not at all. But
+newtypes convert ``case`` into a cast, hence the closure is never
+consumed. So ``newtype`` must not accept non-linear arrow with
+``-XLinearTypes``. These are interpreted as linear ``newtype``-s and a
+warning is emitted (see Specification above).
+
```

Comment:
We believe that this restriction is warranted, because newtypes are intended as a bit-for-bit alias for their contents. Thus, every action performed on the wrapper can be faithfully considered to happen (the same number of times) on the contents.